Контрольная работа 2.1/Контрольная работа 2 задача 1/list.c

```diff
@@ -0,0 +1,176 @@
+#include <stdlib.h>
+#include <stdbool.h>
+#include <stdio.h>
+
+#include "list.h"
+
+typedef struct List
+{
+    int value;
+    List* next;
+} List;
+
+int push(List** head, int value)
+{
+    List* temporary = calloc(1, sizeof(List));
+
+    if (temporary == NULL)
+    {
+        return -1;
+    }
+
+    temporary->value = value;
+    List* node = (*head);
+    List* previous = NULL;
+    if ((*head) == NULL)
+    {
+        temporary->next = (*head);
+        (*head) = temporary;
+        return 0;
+    }
+
+    while (value > node->value)
+    {
+        previous = node;
+        node = node->next;
+        if (node == NULL)
+        {
+            previous->next = temporary;
+            temporary->next = NULL;
+            return 0;
+        }
+    }
+
+    if (previous == NULL)
+    {
+        temporary->next = (*head);
+        (*head) = temporary;
+        return 0;
+    }
+
+    temporary->next = node;
+    previous->next = temporary;
+
+    return 0;
+}
+
+int pop(List** head)
+{
+    if ((*head) == NULL)
+    {
+        return NULL;
+    }
+
+    List* trash = (*head);
+    (*head) = (*head)->next;
+    int temporary = trash->value;
+    free(temporary);
+
+    return trash;
+}
+
+void deleteOddElements(List** head)
+{
+    if ((*head) == NULL || (*head)->next == NULL)
+    {
+        return;
+    }
+    List* temp = NULL;
+    while ((*head) != NULL)
+    {
+        push(&temp, (*head)->value);
+        (*head) = (*head)->next;
+        (*head) = (*head)->next;
+    }
+
+    (*head) = temp;
+}
+
+List* getElement(List* head, size_t index)
+{
+    while (head != NULL && index > 0)
+    {
+        head = head->next;
+        --index;
+    }
+
+    return head;
+}
+
+int deleteElement(List** head, size_t index)
+{
+    if (index == 0)
+    {
+        return pop(head);
+    }
+    List* previous = getElement(*head, index - 1);
+    List* element = previous->next;
+    if (element == NULL)
+    {
+        return NULL;
+    }
+
+    int temporary = element->value;
+    previous->next = element->next;
+
+    free(element);
+    return temporary;
+}
+
+void deleteList(List** head)
+{
+    List* previous = NULL;
+    while ((*head)->next)
+    {
+        previous = (*head);
+        *head = (*head)->next;
+        free(previous);
+    }
+    free(*head);
+}
```

Comment:
```suggestion
    free(*head);
    *head = NULL;
}
```
Иначе у вызывающего останется выглядящий нормальным указатель на уже удалённый список ("висячий указатель").